### PR TITLE
2 implement lookup 2

### PR DESCRIPTION
**Pull request by zancas**
Obviously, the unittests are only stubs at this point, but I have other responsibilities to handle tonight.  I'll flesh them out tomorrow.

### CODE REVIEW

File: dnschain/server.py [hunk 1]
```diff
@@ -26,7 +27,27 @@ def lookup(self, name):
...
+        try:
+            data = json.loads(namecoin_string)
+        except ValueError, e:
+            if namecoin_string.startswith("Not Found: "):
+                print "The name: %s was not found in the database, returning None." % (url_safe_name)
+                return None
+            else:
+                print "Instead of JSON we got this:\n%s\n" % (e,)
+                import sys
+                sys.exit(99) #Fail hard.
```
Comment (taoeffect) [inline dnschain/server.py:45]: Hmm, the library should never quit the app. That is not in its "domain" to do. Let an exception be thrown instead.

File: dnschain/server.py [hunk 2]
```diff
@@ -26,7 +27,27 @@ def lookup(self, name):
 
         @param name: The name to lookup, e.g. 'id/dionyziz'
         """
-
-        # namecoin_string = read_from_dnschain(name)
-        data = json.loads(namecoin_string)
+        #uname = name.encode("utf-8")#Who knows what bytes will be submitted?! Look up dnschain spec.
+        url_safe_name = urllib.quote(name, safe="") #Meant to be a url path _component_.
+        path = "/id/%s" % (url_safe_name,)
```
Comment (taoeffect) [inline dnschain/server.py:32]: The name passed in will be in the format if `id/dionyziz`, it shouldn't be converted into it. Don't worry about sanitization. Just put a comment stating that it's the callers responsibility to sanitize the names.

File: dnschain/server.py [hunk 3]
```diff
@@ -17,6 +17,7 @@ def __init__(self, addr, fingerprint):
 
         self.addr = addr
         self.fingerprint = fingerprint
+        self.connection = httplib.HTTPConnection(addr)
```
Comment (taoeffect) [inline dnschain/server.py:20]: Is this a persistent connection...? Usually a new HTTP request is made each time to the server.

Comment (taoeffect) [inline dnschain/server.py:20]: (It's unlikely to work with a persistent connection, but of course you can test that.)

File: dnschain/server.py (hunk 1 above)
Comment (zancas) [inline dnschain/server.py:45]: Fixed.

File: dnschain/server.py (hunk 2 above)
Comment (zancas) [inline dnschain/server.py:32]: OK, commented per your recommendation.

**Comment by zancas** [comment]
I don't think the connection is persistent (which would explain why the requests work).   I think that the class is poorly named.   I get this impression from this doc:

https://docs.python.org/2/library/httplib.html

and from the integration test cases I included at the bottom of the module.

I'm uncomfortable with the odd(incoherent? (at least I don't understand them well)) names in httplib, I think I'll remove httplib, and use urllib exclusively.   But, I'm going to sleep on it first.

**Comment by zancas** [comment]
Regarding the topic of dependencies, is there an architecture doc I should read to learn policy's with respect to dependencies?    Should I start such a document?   What about using docker, any opinions?

**Comment by taoeffect** [comment]
@zancas wrote:

> Regarding the topic of dependencies, is there an architecture doc I should read to learn policy's with respect to dependencies?

There is not at the moment.

> Should I start such a document?

Probably not worth your time. :P

> What about using docker, any opinions?

Not relevant to this library. I of course have opinions on it (in general, not in reference to this library). I'll add you to slack now since you seem to be seriously contributing. Thanks for your work so far btw, and don't forget to commit and push your changes so that I can see your new updates and review them!

File: dnschain/server.py [hunk 4]
```diff
@@ -26,7 +27,27 @@ def lookup(self, name):
...
+        url_safe_name = urllib.quote(name, safe="") #Meant to be a url path _component_.
+        path = "/id/%s" % (url_safe_name,)
+        self.connection.request("GET", path)
+        namecoin_response = self.connection.getresponse()
+        namecoin_string = namecoin_response.read()
+        try:
+            data = json.loads(namecoin_string)
+        except ValueError, e:
+            if namecoin_string.startswith("Not Found: "):
+                print "The name: %s was not found in the database, returning None." % (url_safe_name)
```
Comment (dionyziz) [inline dnschain/server.py:40]: Our library should never print things. Use `logging`.

File: dnschain/server.py [hunk 5]
```diff
@@ -26,7 +27,27 @@ def lookup(self, name):
...
+                return None
+            else:
+                print "Instead of JSON we got this:\n%s\n" % (e,)
```
Comment (dionyziz) [inline dnschain/server.py:43]: Similar

File: dnschain/server.py [hunk 6]
```diff
@@ -26,7 +27,27 @@ def lookup(self, name):
...
+        self.connection.request("GET", path)
+        namecoin_response = self.connection.getresponse()
+        namecoin_string = namecoin_response.read()
+        try:
+            data = json.loads(namecoin_string)
+        except ValueError, e:
+            if namecoin_string.startswith("Not Found: "):
+                print "The name: %s was not found in the database, returning None." % (url_safe_name)
+                return None
+            else:
```
Comment (dionyziz) [inline dnschain/server.py:42]: Nit: Avoid else-after-return

**Comment by zancas** [comment]
OK.

**Comment by zancas** [comment]
I've been given push access to the okTurtles/pydnschain repo, so the pull request mechanism is obsolete (relative to my commits).

**Comment by dionyziz** [comment]
No, every code change should still be pull requested and merged by someone else, not you, even if you have push access. Let us review you so that we can be sure the code looks good. I do the same with my code :)

**Comment by zancas** [comment]
OK, I understand the policy.

The work flow I was familiar with went like this:
1.  Open an issue on the github repo describing the feature requested, or the bug to be fixed.
2.  Once the issue has been opened that creates an issue number.
3.  Use the issue number mentioned in "1." to name a new branch according to the following format:
       $ISSUE_NUMBER-$DESCRIPTION-OF-BUG-OR-FEATURE_$NUMBER_OF_TIMES_REBASED
4.  Hack along on the issue branch, until one of several things happens:
        1.  production/master is ahead of the branch in a substantive way
        2.  the branch needs to have a cleaner set of commits
5. Create a new branch that's never been shared before that has the same name as the previous branch, but with the $NUMBER_OF_TIMES_REBASED incremented.   Then rebase that new branch on top of production/master.    _Then_ push that branch to a shared repository.  This has nice features:
   1.  The original development history remains (under original branch name, and is never rebased)
   2.  There's never moment when developers working on the same branch are faced with a skewed history, because NO _SHARED_ BRANCH IS EVER REBASED.
   3.  (Maybe not a feature.)  There're no merge commits, when the branch to be merged to master is finally passing review, the merge process (not strictly guaranteed) will be "fast forward" merge.

What do you think?   If I were to follow that scheme, then I'd create a new branch, rebase it on master, squash it to a nice set of intelligible commits, and _then_ push it to github, and open pull request against it.

File: dnschain/server.py (hunk 4 above)
Comment (zancas) [inline dnschain/server.py:40]: Done

File: dnschain/server.py (hunk 6 above)
Comment (zancas) [inline dnschain/server.py:42]: fixed

**Comment by zancas** [comment]
OK, I've addressed the comments in this pull request.   I'm going to:
1. move my proposal about how to share code into an issue
2. close this pull-request without merging, since there's already a different PR for this issue:  https://github.com/okTurtles/pydnschain/pull/5

**Pull request closed without merging**